pkg/proxy/iptables/proxier.go

```diff
@@ -434,126 +510,213 @@ func (proxier *Proxier) syncProxyRules() error {
-				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --set -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), hostChain, protocol, hosts[i]))
+		// Generate the per-endpoint chains.  We do this in multiple passes so we
+		// can group rules together.
+		endpoints := make([]string, 0)
+		endpointChains := make([]utiliptables.Chain, 0)
+		for _, ep := range info.endpoints {
+			endpoints = append(endpoints, ep)
+			endpointChain := servicePortAndEndpointToServiceChain(name, protocol, ep)
+			endpointChains = append(endpointChains, endpointChain)
+
+			// Create the endpoint chain, retaining counters if possible.
+			if chain, ok := existingChains[utiliptables.Chain(endpointChain)]; ok {
+				writeLine(chainsLines, chain)
 			} else {
-				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), protocol, hosts[i]))
+				writeLine(chainsLines, makeChainLine(endpointChain))
 			}
+			activeChains[endpointChain] = true
 		}
 
-		// proxy
-		rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), info.portal.ip.String(), protocol, protocol, info.portal.port, svcChain))
-
-		for _, publicIP := range info.deprecatedPublicIPs {
-			rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"deprecated-PublicIP portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), publicIP, protocol, protocol, info.portal.port, svcChain))
+		// First write session affinity rules, if applicable.
+		if info.sessionAffinityType == api.ServiceAffinityClientIP {
+			for _, endpointChain := range endpointChains {
+				writeLine(rulesLines,
+					"-A", string(svcChain),
+					"-m", "comment", "--comment", name.String(),
+					"-m", "recent", "--name", string(endpointChain),
+					"--rcheck", "--seconds", fmt.Sprintf("%d", info.stickyMaxAgeSeconds), "--reap",
+					"-j", string(endpointChain))
+			}
 		}
 
-		for _, ingress := range info.loadBalancerStatus.Ingress {
-			if ingress.IP != "" {
-				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"load-balancer portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), ingress.IP, protocol, protocol, info.portal.port, svcChain))
+		// Now write loadbalancing & DNAT rules.
+		n := len(endpointChains)
+		for i, endpointChain := range endpointChains {
+			// Balancing rules in the per-service chain.
+			args := []string{
+				"-A", string(svcChain),
+				"-m", "comment", "--comment", name.String(),
 			}
+			if i < (n - 1) {
+				// Each rule is a probabilistic match.
+				args = append(args,
+					"-m", "statistic",
+					"--mode", "random",
+					"--probability", fmt.Sprintf("%f", 1.0/float64(n-i)))
```

Comment:
Done in next push

On Mon, Aug 17, 2015 at 4:30 PM, Abhi Shah <notifications@github.com> wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/kubernetes/kubernetes/pull/12764#discussion_r37249463>
> :
>
> >  			}
> > +			if i < (n - 1) {
> > +				// Each rule is a probabilistic match.
> > +				args = append(args,
> > +					"-m", "statistic",
> > +					"--mode", "random",
> > +					"--probability", fmt.Sprintf("%f", 1.0/float64(n-i)))
>
> (discussed offline)
> we should probably restrict decimal length
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12764/files#r37249463>.
>